core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/progress/ProgressListenerContext.java

```diff
@@ -0,0 +1,162 @@
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.progress;
+
+import software.amazon.awssdk.annotations.Immutable;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.core.SdkResponse;
+import software.amazon.awssdk.core.progress.listener.ProgressListener;
+import software.amazon.awssdk.core.progress.snapshot.ProgressSnapshot;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.http.SdkHttpResponse;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+/**
+ * An SDK-internal implementation of {@link ProgressListener.Context.ExecutionSuccess} and its parent interfaces.
+ * An instance of this class can be used by ProgressListener methods to capture and store request progress
+ * @see ProgressListenerFailedContext for failed transaction context definitions
+ */
+@SdkInternalApi
+@Immutable
+public final class ProgressListenerContext
+    implements ProgressListener.Context.ExecutionSuccess,
+               ToCopyableBuilder<ProgressListenerContext.Builder, ProgressListenerContext> {
+
+    private final SdkRequest request;
+    private final SdkHttpRequest httpRequest;
+    private final ProgressSnapshot uploadProgressSnapshot;
+    private final ProgressSnapshot downloadProgressSnapshot;
+    private final SdkHttpResponse httpResponse;
+    private final SdkResponse response;
+
+    private ProgressListenerContext(Builder builder) {
+        this.request = builder.request;
+        this.httpRequest = builder.httpRequest;
+        this.uploadProgressSnapshot = builder.uploadProgressSnapshot;
+        this.downloadProgressSnapshot = builder.downloadProgressSnapshot;
+        this.httpResponse = builder.httpResponse;
+        this.response = builder.response;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public Builder toBuilder() {
+        return new Builder(this);
+    }
+
+    @Override
+    public SdkRequest request() {
+        return request;
+    }
+
+    @Override
+    public SdkHttpRequest httpRequest() {
+        return httpRequest;
+    }
+
+    @Override
+    public ProgressSnapshot uploadProgressSnapshot() {
+        return uploadProgressSnapshot;
+    }
+
+    @Override
+    public ProgressSnapshot downloadProgressSnapshot() {
+        return downloadProgressSnapshot;
+    }
+
+    @Override
+    public SdkHttpResponse httpResponse() {
+        return httpResponse;
+    }
+
+    @Override
+    public SdkResponse response() {
+        return response;
+    }
+
+    @Override
+    public String toString() {
+        return ToString.builder("ProgressListenerContext")
+                       .add("request", request)
```

Comment:
Can we double-check if logging of request doesnot print any sensitive information related to the request , like authorization header etc

Reply:
Okay, I checked out the docs for SdkPreviewApi 

```
Marker interface for preview and experimental APIs. Breaking changes may be
 * introduced to elements marked as {@link SdkPreviewApi}. Users of the SDK
 * should assume that anything annotated as preview will change or break, and
 * <b>should not</b> use them in production.
 ```

I thought we would get this into pre-prod, test it and then remove it.

Reply:
Once we merge it to Master its as good as Prod.
The project branches merge for projects that will be fully released as released as @SdkPublicApi
So SdkPreviewApi is added only if we plan to release it in master just for users to try it out in preview mode

Reply:
Got it. Incorrect reply for this comment my bad. 

Discussed offline regarding security implications of providing the request object as part of a String. The toString() takes care of omitting private/sensitive information like auth headers, etc.

Reply:
Thanks